client/deposit_ethereum.go

```diff
@@ -0,0 +1,467 @@
+	// check amount
+	if amount < 0 {
+		return CryptoDepositResponse{}, ErrInvalidAmount
+	}
+
+	// check if logged in or not
+	err := c.CheckAuth()
+	if err != nil {
+		return CryptoDepositResponse{}, err
+	}
+
+	// coin status here
+	coinStatus, err := c.getCoinStatusPayload(currency)
+	if err != nil {
+		return CryptoDepositResponse{}, err
+	}
+
+	// getting vault id here
+	vaultID, err := c.getVaultID(ctx, currency)
+	if err != nil {
+		return CryptoDepositResponse{}, err
+	}
+
+	// contract already build in client
+
+	// Getting balance information here
+	balance, err := getTokenBalance(ctx, c.ethClient, ethAddress, currency, coinStatus.BlockchainDecimal, coinStatus.TokenContract)
+	if err != nil {
+		return CryptoDepositResponse{}, err
+	}
+
+	// a more verbose error here could be possible
+	if balance.Cmp(big.NewFloat(amount)) == -1 {
+		return CryptoDepositResponse{}, fmt.Errorf("current balance: %v", balance)
+	}
+
+	var transaction *types.Transaction
+	starkPublicKeyBigInt, ok := new(big.Int).SetString(starkPublicKey[2:], 16)
+	if !ok {
+		return CryptoDepositResponse{}, fmt.Errorf("failed to convert starkPublicKey to big.Int")
+	}
+
+	starkAssetIDBigInt, ok := new(big.Int).SetString(coinStatus.StarkAssetID[2:], 16)
+	if !ok {
+		return CryptoDepositResponse{}, fmt.Errorf("failed to convert StarkAssetID to big.Int")
+	}
+
+	vaultIDBigInt := big.NewInt(int64(vaultID))
+
+	// decimal
+	decimal, err := strconv.Atoi(coinStatus.Decimal)
+	if err != nil {
+		return CryptoDepositResponse{}, err
+	}
+
+	// blockchain decimal
+	blockchainDecimal, err := strconv.Atoi(coinStatus.BlockchainDecimal)
+	if err != nil {
+		return CryptoDepositResponse{}, err
+	}
+
+	// quantization
+	quantization, err := strconv.Atoi(coinStatus.Quanitization)
+	if err != nil {
+		return CryptoDepositResponse{}, nil
+	}
+	quantizedAmount := ToWei(amount, quantization)
+
+	// signer function
+	privateKey, err := crypto.HexToECDSA(ethPrivateKey)
+	if err != nil {
+		return CryptoDepositResponse{}, err
+	}
+
+	signerFn := func(addr common.Address, tx *types.Transaction) (*types.Transaction, error) {
+		// mayhul
+		chainID, err := c.ethClient.ChainID(context.Background())
+		if err != nil {
+			return nil, err
+		}
+
+		signedTx, err := types.SignTx(tx, types.NewLondonSigner(chainID), privateKey)
+		if err != nil {
+			return nil, err
+		}
+
+		return signedTx, nil
+	}
+
+	if currency == ETH {
+		opt, err := getTransactionOpt(ctx, c.ethClient, ethPrivateKey, signerFn, amount, blockchainDecimal)
+		if err != nil {
+			return CryptoDepositResponse{}, err
+		}
+
+		transaction, err = c.starkexContract.DepositEth(opt, starkPublicKeyBigInt, starkAssetIDBigInt, vaultIDBigInt)
+		if err != nil {
+			return CryptoDepositResponse{}, err
+		}
+
+	} else {
+		opt, err := getTransactionOpt(ctx, c.ethClient, ethPrivateKey, signerFn, 0, decimal)
+		if err != nil {
+			return CryptoDepositResponse{}, err
+		}
+		opt.GasLimit = 100000
+
+		if !c.ethereumNetworkAllowanceSet {
```

Comment:
is it possible that this flag is set to true when you set allowance for say btc, but now allowance wont be set when you try depositing some other erc20 token

Reply:
The user can call SetAllwance function to set the allowance and it will work fine but I noticed that this is not go-routine safe, so I will move the logic and rather tell user to call the function manually which will give more controll